controllers/challenge.js

```diff
@@ -0,0 +1,498 @@
+}
+
+exports.returnNextChallenge = function(req, res, next) {
+  if (!req.user) {
+   return res.redirect('../challenges/learn-how-free-code-camp-works');
+  }
+  var completed = req.user.completedChallenges.map(function (elem) {
+    return elem._id;
+  });
+
+  req.user.uncompletedChallenges = resources.allChallengeIds()
+    .filter(function (elem) {
+      if (completed.indexOf(elem) === -1) {
+        return elem;
+      }
+    });
+  // It's not important to wait for the save to finish as the updated user
+  // object is already in memory! We'll just check to see if there was an
+  // error saving and log it out to investigate in our logs.
+  req.user.save(function(err, data) {
+    if (err) {
+      debug('%s saving user!', err);
+    }
+  });
+  // find the user's current challenge and block
+  // look in that block and find the index of their current challenge
+  // if index + 1 < block.challenges.length
+  // serve index + 1 challenge
+  // otherwise increment block key and serve the first challenge in that block
+  // unless the next block is undefined, which means no next block
+  var nextChallengeName;
+
+  var challengeId = String(req.user.currentChallenge.challengeId);
+  var challengeBlock = req.user.currentChallenge.challengeBlock;
+  var indexOfChallenge = challengeMapWithIds[challengeBlock]
+    .indexOf(challengeId);
+
+  if (indexOfChallenge + 1
+    < challengeMapWithIds[challengeBlock].length) {
+    nextChallengeName =
+      challengeMapWithNames[challengeBlock][++indexOfChallenge];
+  } else if (typeof challengeMapWithIds[++challengeBlock] !== 'undefined') {
+    nextChallengeName = R.head(challengeMapWithNames[challengeBlock]);
+  } else {
+    req.flash('errors', {
+      msg: 'It looks like you have finished all of our challenges.' +
+      ' Great job! Now on to helping nonprofits!'
+    });
+    nextChallengeName = R.head(challengeMapWithNames[0].challenges);
+  }
+
+  var nameString = nextChallengeName.trim()
+    .toLowerCase()
+    .replace(/\s/g, '-');
+  return res.redirect('../challenges/' + nameString);
+};
+
+exports.returnCurrentChallenge = function(req, res, next) {
+  if (!req.user) {
+   return res.redirect('../challenges/learn-how-free-code-camp-works');
+  }
+  var completed = req.user.completedChallenges.map(function (elem) {
+    return elem._id;
+  });
+
+  req.user.uncompletedChallenges = resources.allChallengeIds()
+    .filter(function (elem) {
+      if (completed.indexOf(elem) === -1) {
+        return elem;
+      }
+    });
+  req.user.save();
+  if (!req.user.currentChallenge) {
+    req.user.currentChallenge = {};
+    req.user.currentChallenge.challengeId = challengeMapWithIds['0'][0];
+    req.user.currentChallenge.challengeName = challengeMapWithNames['0'][0];
+    req.user.currentChallenge.challengeBlock = '0';
+    req.user.save();
+    return res.redirect('../challenges/learn-how-free-code-camp-works');
+  }
+  var nameString = req.user.currentChallenge.challengeName.trim()
```

Comment:
saving in here is actually unimportant, and could be removed entirely. It's not happening in the callback to allow the next challenge to be served faster and because it's not mission critical.